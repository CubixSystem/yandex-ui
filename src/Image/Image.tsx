import React, {
    useCallback,
    useState,
    useEffect,
    useMemo,
    RefObject,
    ReactNode,
    MouseEventHandler,
    useRef,
} from 'react';

import { cn } from '@bem-react/classname';

import { mergeAllRefs } from '../lib/mergeRefs';

import './Image.css';

export interface IImageProps {
    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLImageElement>;
    /**
     * Ссылка на изображение
     */
    src?: string;
    /**
     * Набор источников для создания [адаптивных изображений](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images). Подробнее об атрибуте `srcSet` в документации [MDN web docs](https://developer.mozilla.org/ru/docs/Web/HTML/Element/img#attr-srcset)
     */
    srcSet?: string;
    /**
     * Набор размеров для создания [адаптивных изображений](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images). Подробнее об атрибуте `sizes` в документации [MDN web docs](https://developer.mozilla.org/ru/docs/Web/HTML/Element/img#attr-sizes)
     */
    sizes?: string;
    /**
     * URL картинки для экранов с высокой плотностью пикселей = srcSet: "url 2x". srcSet имеет приоритет.
     */
    src2x?: string;
    /**
     * Ссылка на изображение при неудачной загрузке
     */
    fallbackSrc?: string;
    /**
     * Дополнительный класс
     */
    className?: string;
    /**
     * Дополнительный класс для изображения
     */
    imageClassName?: string;
    /**
     * Компонент для отображения во время загрузки картинки
     */
    stub?: ReactNode;
    /**
     * Альтернативный текст
     */
    alt?: string;
    /**
     * атрибут aria-label
     */
    ariaLabel?: string;
    /**
     * обработчик клика
     */
    onClick?: MouseEventHandler<HTMLElement>;
    /**
     * атрибут [loading у изображения](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-loading)
     */
    loading?: 'lazy' | 'eager' | 'auto';
    /**
     * Ширина изображения
     */
    width?: number | string;
    /**
     * Высота изображения
     */
    height?: number | string;
    /**
     * Скругления изображения
     */
    borderRadius?: number | string;
}

const cnImage = cn('Image');

const emptyImage = 'data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=';

const ANIMATION_DELAY_MS = 50;

/**
 * Компонент для красивой загрузки картинок

 * - Пока загружается картинка, пользователь будет видеть `<stub/>`
 * - После загрузки картинки она будет плавно отображена поверх `<stub/>`
 * - После плавного показана картинки `<stub/>` будет удалён из `DOM`
 * - Если картинка загружается из кэша, то пользователь увидит сразу картинку без показа `<stub/>`
 * - Если картинку не удалось загрузить, то покажется стандартная картинка из свойства `fallback`
 * @param {IImageProps} props
 */
export const Image: React.FC<IImageProps> = (props) => {
    const {
        src,
        src2x,
        srcSet,
        sizes,
        fallbackSrc,
        className,
        imageClassName,
        stub,
        alt,
        ariaLabel,
        onClick,
        loading = 'lazy',
        width,
        height,
        borderRadius,
        innerRef,
    } = props;

    const [isLoaded, setLoaded] = useState(false);
    const [isFailed, setFailed] = useState(false);
    const [needAnimation, setNeedAnimation] = useState(false);
    const [canRemoveStub, setCanRemoveStub] = useState(false);

    // нужен реф по условию, а значит useRef не годится
    // важный момент: в innerRef должна быть актуальная ссылка на ref
    const imageRef = useRef<HTMLImageElement>(null);

    // обьединяем рефы
    const mergedRefs = useMemo(() => mergeAllRefs<HTMLImageElement>(imageRef, innerRef), [imageRef, innerRef]);

    const imageSrc = src || emptyImage;
    const legacySrcSet = src2x ? `${src2x} 2x` : undefined;
    const imageSrcSet = srcSet || legacySrcSet;

    const isRendered = useCallback(() => {
        if (!imageRef.current) {
            return false;
        }
        const { naturalWidth, naturalHeight } = imageRef.current;

        return naturalWidth > 0 && naturalHeight > 0;
    }, [imageRef]);

    useEffect(() => {
        const timer = setTimeout(() => {
            if (!isLoaded && !isRendered()) {
                setNeedAnimation(true);
            }
        }, ANIMATION_DELAY_MS);

        return () => clearTimeout(timer);
    }, [isLoaded, isRendered, setNeedAnimation]);

    const handleLoad = useCallback(() => {
        setLoaded(true);
    }, []);

    const handleError = useCallback(() => {
        setFailed(true);
    }, []);
    const containerClassNames = cnImage('Container', [className]);
    const imageClassNames = cnImage(
        {
            loaded: isLoaded,
            loading: needAnimation,
            animated: needAnimation && isLoaded,
        },
        // Если нет обертки, то добавляем className в изображение
        [!stub ? className : undefined, imageClassName],
    );

    const handleOnAnimationEnd = useCallback(() => {
        setCanRemoveStub(true);
    }, []);

    const sizeAttrs: Pick<IImageProps, 'width' | 'height' | 'borderRadius'> = {};
    // безопасно устанавливаем стили, чтобы избежать undefined в html
    if (width !== undefined) {
        sizeAttrs.width = width;
    }
    if (height !== undefined) {
        sizeAttrs.height = height;
    }
    if (borderRadius !== undefined) {
        sizeAttrs.borderRadius = borderRadius;
    }

    const image = (
        <img
            style={sizeAttrs}
            alt={alt}
            aria-hidden={ariaLabel ? 'false' : 'true'}
            aria-label={ariaLabel}
            className={imageClassNames}
            onAnimationEnd={handleOnAnimationEnd}
            onClick={onClick}
            onError={handleError}
            onLoad={handleLoad}
            ref={mergedRefs}
            src={isFailed && fallbackSrc ? fallbackSrc : imageSrc}
            srcSet={imageSrcSet}
            sizes={sizes}
            // TypeScript пока не знает про такой атрибут
            // @ts-ignore
            loading={loading}
        />
    );

    if (stub) {
        return (
            <div className={containerClassNames} style={sizeAttrs}>
                {needAnimation && !canRemoveStub && stub}
                {image}
            </div>
        );
    }

    return image;
};

Image.displayName = cnImage();
