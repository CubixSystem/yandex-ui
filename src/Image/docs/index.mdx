---
id: Image
path: /components/image

title: Image
description: Компонент для красивой загрузки картинок.

images:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Image
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Link.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=1473%3A9880
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fcontent-image-desktop--playground

tabs:
 - Examples
 - Code
---
