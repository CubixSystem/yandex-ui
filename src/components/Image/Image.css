.Image-Container {
    position: relative;
}

.Image {
    visibility: hidden; /* Позволяет не показывать битые картинки при ошибках загрузки */

    font: 0/0 a;

    color: #fff;
    border: 0;
}

.Image_loaded {
    visibility: visible;
}

.Image_loading {
    opacity: 0;
}

.Image_animated {
    animation: Image_fade_in .4s ease forwards;
}

@keyframes Image_fade_in {
    from {
        opacity: 0;
    }

    to {
        opacity: 1;
    }
}
