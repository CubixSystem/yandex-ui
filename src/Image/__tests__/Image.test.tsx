import React, { createRef } from 'react';
import {
    createClientRender,
    screen,
    fireEvent,
    act,
    createServerRender,
    createContainer,
} from '../../internal/testing';
import { Image } from '../Image';

const EMPTY_IMAGE_SRC = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';
const SECOND_EMPTY_IMAGE_SRC =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQMAAAAl21bKAAAAA1BMVEUAAACnej3aAAAAAXRSTlMAQObYZgAAAApJREFUCNdjYAAAAAIAAeIhvDMAAAAASUVORK5CYII=';

describe('Image', () => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(<Image src={EMPTY_IMAGE_SRC} className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен вернуть полный шаблон компонента со стабов (snapshot)', () => {
        const { container } = render(<Image src={EMPTY_IMAGE_SRC} stub={<div className="stub" />} className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен вернуть полный шаблон компонента с пустым изображением', () => {
        const { container } = render(<Image src="" className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('Должен быть установлен корректный displayName', () => {
        expect(Image.displayName).toBe('Image');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', async () => {
        const innerRef = createRef<HTMLImageElement>();
        render(<Image innerRef={innerRef} src={EMPTY_IMAGE_SRC} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен установить атрибут src при наличии свойства src', () => {
        render(<Image src={EMPTY_IMAGE_SRC} />);

        expect(screen.getByRole('img', { hidden: true })).toHaveAttribute('src', EMPTY_IMAGE_SRC);
    });

    test('должен установить атрибут alt при наличии свойства alt', () => {
        render(<Image alt="Иконка Серпа" src={EMPTY_IMAGE_SRC} />);

        expect(screen.getByRole('img', { hidden: true })).toHaveAttribute('alt', 'Иконка Серпа');
    });

    test('должен установить атрибут width и height при наличии свойства width и height', () => {
        render(<Image width={200} height={200} src={EMPTY_IMAGE_SRC} />);

        expect(screen.getByRole('img', { hidden: true })).toHaveStyle({ height: '200px', width: '200px' });
    });

    test('должен установить атрибут srcset и sizes при наличии свойства srcset и sizes', () => {
        render(<Image srcSet={`${EMPTY_IMAGE_SRC} 2x`} sizes="(max-width: 600px) 200px, 50vw" src={EMPTY_IMAGE_SRC} />);

        expect(screen.getByRole('img', { hidden: true })).toHaveAttribute('srcSet', `${EMPTY_IMAGE_SRC} 2x`);
        expect(screen.getByRole('img', { hidden: true })).toHaveAttribute('sizes', '(max-width: 600px) 200px, 50vw');
    });

    test('подставляется fallback картинка', () => {
        render(<Image src={EMPTY_IMAGE_SRC} fallbackSrc={SECOND_EMPTY_IMAGE_SRC} />);

        fireEvent.error(screen.getByRole('img', { hidden: true }));

        expect(screen.getByRole('img', { hidden: true })).toHaveAttribute('src', SECOND_EMPTY_IMAGE_SRC);
    });

    test('должен появиться стаб', () => {
        jest.useFakeTimers();

        render(<Image stub={<div data-testid="stub" />} src={EMPTY_IMAGE_SRC} />);

        act(() => {
            jest.runAllTimers();
        });

        expect(screen.getByTestId('stub')).toBeInTheDocument();

        jest.useRealTimers();
    });

    test('должен исчезнуть стаб после окончания анимации', () => {
        jest.useFakeTimers();

        render(<Image stub={<div data-testid="stub" />} src={EMPTY_IMAGE_SRC} />);

        act(() => {
            jest.runAllTimers();
        });

        fireEvent.animationEnd(screen.getByRole('img', { hidden: true }));

        expect(() => screen.getByTestId('stub')).toThrow();

        jest.useRealTimers();
    });

    test('стаб не должен появиться если изображение загружено', () => {
        jest.useFakeTimers();

        render(<Image stub={<div data-testid="stub" />} src={EMPTY_IMAGE_SRC} />);

        fireEvent.load(screen.getByRole('img', { hidden: true }));

        act(() => {
            jest.runAllTimers();
        });

        expect(() => screen.getByTestId('stub')).toThrow();

        jest.useRealTimers();
    });

    test('должен отображаться после гидрации', async () => {
        const Fixture = () => <Image />;

        const { html } = createServerRender()(<Fixture />);
        const container = createContainer(html);
        document.body.appendChild(container);

        await new Promise((resolve) => setTimeout(resolve, 50));

        // проверяем что изображение загружено до гидрации
        expect(document.body.getElementsByTagName('img')[0].complete).toBe(true);

        render(<Fixture />, { container, hydrate: true });

        expect(screen.getByRole('img', { hidden: true })).toHaveClass('Image_loaded');
    });
});
