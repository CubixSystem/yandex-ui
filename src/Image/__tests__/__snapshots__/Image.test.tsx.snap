// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image должен вернуть полный шаблон компонента (snapshot) 1`] = `
<img
  aria-hidden="true"
  class="Image mix"
  loading="lazy"
  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
/>
`;

exports[`Image должен вернуть полный шаблон компонента с пустым изображением 1`] = `
<img
  aria-hidden="true"
  class="Image mix"
  loading="lazy"
  src="data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs="
/>
`;

exports[`Image должен вернуть полный шаблон компонента со стабов (snapshot) 1`] = `
<div
  class="Image-Container mix"
>
  <img
    aria-hidden="true"
    class="Image"
    loading="lazy"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
  />
</div>
`;
